### PR TITLE
Fix HasError and HasLocks

### PR DESCRIPTION
**Pull request by ekexium**
Related issues: #242 #238 

The PR tries to fix the following issues:
1. `HasError for Result` ignores `Err`
2. `HasLocks` ignores the `error` field
3. If a request returns `Ok(response)` while the response contains a non-empty `error` field, it's often ignored.

### CODE REVIEW

File: src/transaction/requests.rs [hunk 1]
```diff
@@ -547,13 +563,82 @@ pub struct SecondaryLocksStatus {
     pub commit_ts: Option<Timestamp>,
 }
 
-impl HasLocks for kvrpcpb::CommitResponse {}
+impl HasLocks for kvrpcpb::CommitResponse {
```
Comment (nrc) [inline src/transaction/requests.rs:566]: Could you use a macro for these impls, it's a lot of duplicated code.

File: src/request/plan.rs [hunk 2]
```diff
@@ -249,6 +249,35 @@ where
     }
 }
 
+pub struct PropagateError<P: Plan> {
```
Comment (nrc) [inline src/request/plan.rs:252]: Could you document the plan please? I also think the name is not quite right, maybe `ExtractError` or something is better?

**Review by nrc — COMMENTED** [review]
Could you add some tests please? Otherwise its looking pretty good.

File: src/request/plan.rs (hunk 2 above)
Comment (ekexium) [reply src/request/plan.rs:252]: My first thought was `ExtractError` too, but there exists a similar name `extract_errors`.

File: src/transaction/requests.rs (hunk 1 above)
Comment (ekexium) [reply src/transaction/requests.rs:566]: Sure

File: src/request/plan.rs (hunk 2 above)
Comment (nrc) [reply src/request/plan.rs:252]: It's in a pretty different part of the system, so I think it should be ok

**Comment by ekexium** [comment]
Comments addressed, PTAL @nrc

**Review by nrc — APPROVED** [review]
lgtm!

**Pull request merged**